### PR TITLE
Improve S.IO.Ports

### PR DESCRIPTION
**Pull request by josesimoes**
## Description
- Update native code to handle most serial tasks in native code.
- Update S.IO.Ports assembly declaration.

## Motivation and Context
- General performance improvement.
- Following nanoframework/System.IO.Ports#32.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [x] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp [hunk 1]
```diff
@@ -276,816 +286,844 @@ HRESULT Library_sys_io_ser_native_System_IO_Ports_SerialPort::get_BytesToRead___
...
+        }
 
-    // Check RS485 mode is not selected as currently not supported
-    if ((SerialMode)pThis[FIELD___mode].NumericByRef().s4 != SerialMode_Normal)
+        // fill data buffer from Rx buffer
+        palUart->RxRingBuffer.Pop(buffer, bufferLength);
+
+        NANOCLR_CHECK_HRESULT(CLR_RT_HeapBlock_String::CreateInstance(top, (const char *)buffer, bufferLength));
+
+        platform_free(buffer);
```
Comment (Ellerbach) [inline targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp:475]: You have a risk here that the buffer won't be free if the previous check fails.

File: targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp [hunk 2]
```diff
@@ -276,816 +286,844 @@ HRESULT Library_sys_io_ser_native_System_IO_Ports_SerialPort::get_BytesToRead___
...
+    // pop "hbTimeout" heap block from stack
+    stack.PopValue();
+
+    // return how many bytes were read
+    stack.SetResult_String((const char *)line);
+
+    // free memory, if needed
+    if (line != NULL)
+    {
+        platform_free(line);
```
Comment (Ellerbach) [inline targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp:574]: Same risk here as you have checks before. So just raising this to make sure it's not a possibility and that the line will be freed whatever happens before

File: targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp [hunk 3]
```diff
@@ -276,816 +286,844 @@ HRESULT Library_sys_io_ser_native_System_IO_Ports_SerialPort::get_BytesToRead___
...
+
+    // pop "hbTimeout" heap block from stack
+    stack.PopValue();
+
+    stack.SetResult_U4(length);
+
+    // free memory, if it was allocated
+    if (isNewAllocation && buffer)
+    {
+        platform_free(buffer);
```
Comment (Ellerbach) [inline targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp:1122]: And same question about the risk

File: targets/FreeRTOS/NXP/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp [hunk 4]
```diff
@@ -177,536 +182,656 @@ HRESULT Library_sys_io_ser_native_System_IO_Ports_SerialPort::get_BytesToRead___
...
-        }
+    // return how many bytes were read
+    stack.SetResult_String((const char *)line);
 
-        // null pointers and vars
-        pThis = NULL;
+    // free memory, if needed
+    if (line != NULL)
+    {
+        platform_free(line);
```
Comment (Ellerbach) [inline targets/FreeRTOS/NXP/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp:481]: And one more similar question here as well

File: targets/FreeRTOS/NXP/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp [hunk 5]
```diff
@@ -177,536 +182,656 @@ HRESULT Library_sys_io_ser_native_System_IO_Ports_SerialPort::get_BytesToRead___
...
-        base->CTRL |= 1U << LPUART_CTRL_ORIE_SHIFT;
-        // Renable transmitter and receiver
-        base->CTRL |= 1U << 19;
-        base->CTRL |= 1U << 18;
+        // fill data buffer from Rx buffer
+        palUart->RxRingBuffer.Pop(buffer, bufferLength);
+
+        NANOCLR_CHECK_HRESULT(CLR_RT_HeapBlock_String::CreateInstance(top, (const char *)buffer, bufferLength));
+
+        platform_free(buffer);
```
Comment (Ellerbach) [inline targets/FreeRTOS/NXP/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp:377]: I guess I won't ask the question everywhere :-) There are something like 15+ of those :-) Just want to make sure whatever happens the buffer is properly freed

**Review by Ellerbach — APPROVED** [review]
WOW. Amazing work and refactoring. Very impressive. Only questions I have is regarding the platform_free, just to make sure they always happen. Reason is because there is some code checking for HRESULT and redirecting to the end of the function. And if the bugger has been allocated before and for some reason that check will fail, then it will never be deallocated.
Other than that, rest looks excellent!

File: targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp (hunk 1 above)
Comment (josesimoes) [reply targets/ChibiOS/_nanoCLR/System.IO.Ports/sys_io_ser_native_System_IO_Ports_SerialPort.cpp:475]: Good catch on this one! I did try to make those calls to `platform_free` to always be on the exit path. 😅 
Sometimes it does look that's been missed, but that's when the buffer hasn't been allocated.
I'll double check these again and add comments for those cases.

**Pull request merged**